consensus/qbft/src/main/java/org/hyperledger/besu/consensus/qbft/pki/PkiQbftExtraDataCodec.java

```diff
@@ -0,0 +1,124 @@
+import org.hyperledger.besu.ethereum.rlp.RLPInput;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+
+import org.apache.tuweni.bytes.Bytes;
+
+public class PkiQbftExtraDataCodec extends QbftExtraDataCodec {
+
+  @Override
+  public BftExtraData decodeRaw(final Bytes input) {
+    if (input.isEmpty()) {
+      throw new IllegalArgumentException("Invalid Bytes supplied - Bft Extra Data required.");
+    }
+
+    final RLPInput rlpInput = new BytesValueRLPInput(input, false);
+
+    rlpInput.enterList(); // This accounts for the "root node" which contains BFT data items.
+    final Bytes vanityData = rlpInput.readBytes();
+    final List<Address> validators = rlpInput.readList(Address::readFrom);
+
+    final Optional<Vote> vote;
+    if (rlpInput.nextIsList() && rlpInput.nextSize() == 0) {
+      vote = Optional.empty();
+      rlpInput.skipNext();
+    } else {
+      vote = Optional.of(decodeVote(rlpInput));
+    }
+
+    final int round = rlpInput.readIntScalar();
+    final List<SECPSignature> seals =
+        rlpInput.readList(
+            rlp -> SignatureAlgorithmFactory.getInstance().decodeSignature(rlp.readBytes()));
+
+    final Optional<Bytes> cms;
+    if (rlpInput.nextIsList() && rlpInput.nextSize() == 0) {
+      cms = Optional.empty();
+      rlpInput.skipNext();
+    } else {
+      rlpInput.enterList();
+      cms = Optional.of(rlpInput.readBytes());
+      rlpInput.leaveList();
+    }
+
+    rlpInput.leaveList();
+
+    return new PkiQbftExtraData(vanityData, seals, vote, round, validators, cms);
+  }
+
+  @Override
+  protected Bytes encode(final BftExtraData bftExtraData, final EncodingType encodingType) {
+    if (!(bftExtraData instanceof PkiQbftExtraData)) {
+      throw new IllegalStateException(
+          "PkiQbftExtraDataCodec must be used only with PkiQbftExtraData");
+    }
+    final PkiQbftExtraData extraData = (PkiQbftExtraData) bftExtraData;
+
+    final BytesValueRLPOutput encoder = new BytesValueRLPOutput();
+    encoder.startList();
+    encoder.writeBytes(extraData.getVanityData());
+    encoder.writeList(extraData.getValidators(), (validator, rlp) -> rlp.writeBytes(validator));
+
+    if (extraData.getVote().isPresent()) {
+      encodeVote(encoder, extraData.getVote().get());
+    } else {
+      encoder.writeList(Collections.emptyList(), (o, rlpOutput) -> {});
+    }
+
+    if (encodingType != EncodingType.EXCLUDE_COMMIT_SEALS_AND_ROUND_NUMBER) {
+      encoder.writeIntScalar(extraData.getRound());
+      if (encodingType != EncodingType.EXCLUDE_COMMIT_SEALS) {
+        encoder.writeList(
+            extraData.getSeals(), (committer, rlp) -> rlp.writeBytes(committer.encodedBytes()));
+      } else {
+        encoder.writeEmptyList();
+      }
+    } else {
+      encoder.writeIntScalar(0);
+      encoder.writeEmptyList();
+    }
+
+    if (encodingType == EncodingType.ALL) {
+      if (extraData.getCms().isPresent()) {
```

Comment:
Looks like the cms is excluded when calculating the onchain block hash and committed seal? Seems like you probably always want the cms to be included as part of the hash?

Reply:
We can't have the CMS as part of the on-chain hash calculation because it has the block hash inside of it as a signed payload.